### PR TITLE
NEW: VAT report - Optimisation & collapse by rate

### PR DESCRIPTION
**Pull request by aspangaro**
Thanks to @lvessiller-opendsi 

Step 1 :
For optimization issues for some customers who have large amounts of customer/vendor invoices (otherwise the page is unreachable). Do not load a full year's data by default.

Use the VAT frequency defined in the configuration of the VAT module to calculate the default periodicity when arriving in compta/tva/index.php

If frequency "month", only load default period M-1. Example: January 2021 / Loading period December 2020.

Step 2 :
On the detailed report (quadri_details), make a summary page with "Details" links next to each VAT code for a better visibility.

image

### CODE REVIEW

File: htdocs/core/lib/tax.lib.php [hunk 1]
```diff
@@ -599,16 +599,14 @@ function tax_by_rate($type, $db, $y, $q, $date_start, $date_end, $modetax, $dire
 		$sql .= " f.".$invoicefieldref." as facnum, f.type, f.total_ttc as ftotal_ttc, f.datef, s.nom as company_name, s.rowid as company_id,";
 		$sql .= " p.rowid as pid, p.ref as pref, p.fk_product_type as ptype,";
 		$sql .= " 0 as payment_id, 0 as payment_amount";
-		$sql .= " FROM ".MAIN_DB_PREFIX.$invoicetable." as f,";
-		$sql .= " ".MAIN_DB_PREFIX."societe as s,";
-		$sql .= " ".MAIN_DB_PREFIX.$invoicedettable." as d";
+		$sql .= " FROM ".MAIN_DB_PREFIX.$invoicetable." as f";
+		$sql .= " LEFT JOIN ".MAIN_DB_PREFIX."societe as s ON s.rowid = f.fk_soc";
```
Comment (eldy) [inline htdocs/core/lib/tax.lib.php:603]: Why using a left join instead of default join ?

Comment (aspangaro) [reply htdocs/core/lib/tax.lib.php:603]: A left join is normally faster than a default join. This is optimization. Do you want us to leave the default join?

Comment (eldy) [reply htdocs/core/lib/tax.lib.php:603]: No, this is the opposite. A LEFT JOIN is nearly often slower that an INNER JOIN. On a left join, you say to the database, I want the record on second table even if it does not exists. It result in the execution plan of the database (you can get it by doing "EXPLAIN PLAN sqlrequest"), that the database will not try to use matching on index on both side, but will often make a full scan (on the right or on the left side, this is very very slow).
When you are sure there is a child (due to context rules), you must always use a strict join. Left join must be used only when a child may not exists and we want, despite of this, the parent (this is a different need and result, and it is slower).

Comment (aspangaro) [reply htdocs/core/lib/tax.lib.php]: @eldy Done. Thanks for the explanation.

**Comment by aspangaro** [comment]
Ping @eldy Can you merge this functionnality ?

File: htdocs/compta/tva/index.php [hunk 2]
```diff
@@ -43,61 +44,89 @@
...
-if (empty($date_start) || empty($date_end)) { // We define date_start and date_end
-	$q = GETPOST("q", "int");
-	if (empty($q)) {
-		if (GETPOST("month", "int")) {
-			$date_start = dol_get_first_day($year_start, GETPOST("month", "int"), false);
-			$date_end = dol_get_last_day($year_start, GETPOST("month", "int"), false);
+$refresh = GETPOSTISSET('submit') ? true : false;
+
+if ($refresh === false) {
+	$year_current = intval(strftime('%Y', $now));
```
Comment (eldy) [inline htdocs/compta/tva/index.php:50]: strftime is forbidden (TZ difference between server and user are not correctly managed). Can you replace with dol_print_date($now, '%Y')

Comment (aspangaro) [reply htdocs/compta/tva/index.php]: Noted

File: htdocs/compta/tva/quadri_detail.php [hunk 3]
```diff
@@ -428,7 +427,10 @@
 		if (is_array($x_both[$rate]['coll']['detail'])) {
 			// VAT Rate
 			print "<tr>";
-			print '<td class="tax_rate">'.$langs->trans("Rate").': '.vatrate($rate).'%</td><td colspan="'.($span + 1).'"></td>';
+			print '<td class="tax_rate" colspan="' . ($span+1) . '">';
+			print $langs->trans('Rate') . ' : ' . vatrate($rate) . '%';
+			print ' - <a href="' . dol_buildpath('/compta/tva/quadri_detail.php', 1) . '?invoice_type=customer&amp;vat_rate_show=' . $rate . '&amp;year=' . $year_start . '&amp;month=' . $month_start . '">' . img_picto('', 'chevron-down', 'class="paddingrightonly"') . $langs->trans('VATReportShowByRateDetails') . '</a>';
```
Comment (eldy) [inline htdocs/compta/tva/quadri_detail.php:432]: As this is an internal url, the dol_buildpath (designed for link to external url of modules) can be replaced with
DOL_URL_ROOT.'/compta/tva/quadri_detail.php'
Also any param into an url must be url encode, for example
&amp;vat_rate_show=' .urlencode($rate)

File: htdocs/compta/tva/index.php [hunk 4]
```diff
@@ -43,61 +44,89 @@
...
-			$date_end = dol_get_last_day($year_start, 12, false);
+		$date_start = dol_get_first_day($year, $month_last);
+		$date_end = dol_get_last_day($year, $month_last);
+	}
+} else {
+	// Date range
+	$year = GETPOST("year", "int");
+	if (empty($year)) {
+		$year_current = dol_print_date("%Y", dol_now());
+		if ($conf->global->SOCIETE_FISCAL_MONTH_START > date('m')) $year_current--;
```
Comment (eldy) [inline htdocs/compta/tva/index.php:95]: Same "date" function should not be used to get month, use instead 
dol_print_date(dol_now(), "%m")

File: htdocs/compta/tva/index.php [hunk 5]
```diff
@@ -43,61 +44,89 @@
...
-			$date_start = dol_get_first_day($year_start, 10, false);
-			$date_end = dol_get_last_day($year_start, 12, false);
+		$date_start = dol_get_first_day($year, $month_last);
+		$date_end = dol_get_last_day($year, $month_last);
+	}
+} else {
+	// Date range
+	$year = GETPOST("year", "int");
+	if (empty($year)) {
+		$year_current = dol_print_date("%Y", dol_now());
```
Comment (eldy) [inline htdocs/compta/tva/index.php:94]: Order of parameters seems wrong

**Comment by aspangaro** [comment]
Thanks @eldy. Mistakes are resolved

**Comment by Veance** [comment]
Hi, 

Is it possible to disable this new function (collapse by rate) ? 
via Option or parameter
Thanks, 
regards,

**Comment by lvessiller-opendsi** [comment]
@Veance 

Sorry, but there is no option to disable this collapse by rate.
It would be interesting to add this option if you have'nt a lot of data.
Could you open an Issue for this ?

Thanks,

**Pull request merged**